crypto/bio/bio_lib.c

```diff
@@ -533,9 +533,7 @@ long BIO_ctrl(BIO *b, int cmd, long larg, void *parg)
     return ret;
 }
 
-long BIO_callback_ctrl(BIO *b, int cmd,
-                       void (*fp) (struct bio_st *, int, const char *, int,
-                                   long, long))
+long BIO_callback_ctrl(BIO *b, int cmd, bio_info_cb *fp)
```

Comment:
bio_info_cb should be BIO_info_cb.  Lots of places to change.

Reply:
The name "bio_info_cb" is in a public header, and it is also mentioned in the doc.
Should I still change the name to BIO_info_cb?

Reply:
We can have both.  Create `BIO_info_cb` and keep the definition of `bio_info_cb` for compatibility.  Something like this?

```
typedef void BIO_info_cb(BIO *, int, const char *, int, long, long);
typedef BIO_info_cb bio_info_cb;   /* backward compatibility */
```

Reply:
yes, sounds good.

Reply:
??? `bio_info_cb *fp` is *different* from original `void (*fp)()`, it's extra dereference level. This results in binary incompatibility. `bio_info_cb fp` is equivalent one... Indeed
```
typedef void (*callback)();
void foo(callback *fp) { return (*fp)(); }
void bar(callback fp) { return (*fp)(); }
```
foo is compiled as `call *(%rdi)` while bar as `call *rdi`.

Reply:
An extra level in indirection is of course not intended.

Reply:
Hmmm?  With this definition (currently present in `include/openssl/bio.h`), I don't quite understand how `bio_info_cb *fp` becomes an extra level of indirection...

```
typedef void bio_info_cb(BIO *, int, const char *, int, long, long);
```

Reply:
no you are right, there is no extra level of indirection,
but if you look at the .pod one could think so.

Reply:
Ah!  Fix docs, please :-)

Reply:
(or well, I can...  actually, I'm thinking we might want to have docs-nits check that, but that's a rather complex matter)

Reply:
PR created...

Reply:
> there is no extra level of indirection,

Right, I apologize for falsehood. It's just that I'm so accustomed to `typedef (*callback)()` that I haven't even considered possibility of `typedef callback()`. And I'd argue that the fact that .pod was incorrect is indication of the fact that former is considered more common/intuitive/readable...

Reply:
Another thing that tips me off in this function, BIO_callback_ctrl, is (void *)&fp. It's meant to be used as `const char *argp` by callee, and what does &fp mean?

Reply:
OK, it's meant to be used as argument to BIO_CB_CTRL, but here is question, would *sane* BIO_CB_CTRL do double indirection?

Reply:
That is the address of "fp".
You can't cast a function pointer to (void *) that triggers another warning.

Reply:
side note: I disagree that the function signature typedef would be less intuitive than a function pointer typedef...  quite the contrary, actually, but that goes beyond just functions, I regard typedefs to get a pointer type evil unless made clear in the name of the new type.

side note two: I know that I haven't acted along my own line of thought.  I'll claim temporary insanity ;-)

Reply:
>> what does &fp mean?

> That is the address of "fp".

I've expressed myself poorly (as usual). It's not like I don't understand what &fp means, question was rather does callee actually recognize that it has to dereference it? But I'm having trouble finding a callee...

Reply:
look at BIO_debug_callback in the crypto/bio/bio_cb.c
at least this one does not.
what makes it worse is that BIO_ctrl does this:
`ret = bio_call_callback(b, BIO_CB_CTRL, parg, 0, cmd, larg, 1L, NULL);`